File: src/main/scala/chicken/cli/commands/Post.scala
package chicken.cli.commands

import chicken.cli.App
import chicken.domain.User

case class Post() extends CliCommand {

  def token: String = " -> "

  override def validate(input: String): List[String] = {
    val trimSplit = App.trimSplit(input, token)
    if (trimSplit.length != 2 ||
      trimSplit(0) == null || trimSplit(0).isEmpty ||
      trimSplit(1) == null || trimSplit(1).isEmpty)
      "Invalid 'post' format. Type 'Username -> message'" :: Nil
    else
      Nil
  }

  def exec(input: String): Unit = {
    val trimSplit = App.trimSplit(input, token)
    val name = trimSplit(0)
    val message = trimSplit(1)
    if (!App.rep.isUserExist(name))
      App.rep.saveUser(User(name))
    App.rep.savePost(chicken.domain.Post(message, name, System.currentTimeMillis()))
  }

}

File: src/chicken/cli/commands/post.clj
(ns chicken.cli.commands.post
  (:require [chicken.domain]
            [chicken.repository :as rep]
            [clojure.string :refer (split blank?)]
            [chicken.cli.app :refer (pluginCommand trim-split)])
  (:import [chicken.domain User Post]))

(def ^:private TOKEN " -> ")

(defn validate [input]
  (let
    [[name message] (trim-split input TOKEN)]
    (if (or (blank? name) (blank? message))
      {:errors '("Invalid 'post' format. Type 'Username -> message'")}
      :ok)))

(defn post [input]
  "Save post from input.
  If username from input does not exist, save username."
  (let
    [[name message] (trim-split input TOKEN)]
    (when-not (rep/existsUser? name)
      (rep/saveUser (User. name #{})))
    (rep/savePost (Post. message name (System/currentTimeMillis)))))

(pluginCommand :token TOKEN :do post :validate validate)
